Step4 리뷰 부탁드립니다.

5minho opened the pull request:
안녕하세요 리뷰어님
이번 단계에서는 세션을 어디서 어떻게 생성할 건지, 어떻게 테스트 할건지 등등 이런저런 고민이 많았습니다.
리뷰 부탁드리겠습니다.
감사합니다.

ohminho-woowahan committed:
feat
- 세션 기본 api 구현

ohminho-woowahan committed:
feat
- 세션 저장소 구현

ohminho-woowahan committed:
refactor
- 세션을 다루는 부분을 HttpSessionHandler 클래스로 이관

ohminho-woowahan committed:
feat
- 로그인 부분을 세션으로 수정

ohminho-woowahan committed:
chore
- 세션핸들러 변수명 수정

sah3122 commented on src/main/java/webserver/HttpSessionStore.java:
```diff
@@ -0,0 +1,10 @@
+
+import webserver.http.HttpSession;
+
+public interface HttpSessionStore {
```
세션 스토어 인터페이스 정의 👍

sah3122 commented on src/main/java/webserver/IdGenerator.java:
```diff
@@ -0,0 +1,7 @@
+package webserver;
+
+interface IdGenerator {
```
ID 생성기도 인터페이스를 분리 해주셨네요 👍 
많은 고민을 하신것이 느껴집니다 !

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -55,6 +57,8 @@ public void run() {
     }
 
     private void handleRequest(HttpRequest httpRequest, HttpResponse httpResponse) throws IOException {
+        sessionHandler.handleSession(httpRequest, httpResponse);
```
모든 요청마다 session handling 과정을 거치도록 구현해주셨네요 💯 
세션을 사용하는 시점에 초기화 하는건 어떻게 생각하시나요 ?

5minho replied:
세션을 사용하는 시점에 초기화가 된다면, 서버에서 불필요한 세션을 생성하지 않아 좋을거 같습니다!

sah3122 commented on src/main/java/webserver/HttpSessionHandler.java:
```diff
@@ -0,0 +1,42 @@
+        return sessionStore.getSession(sessionIdFromCookie);
+    }
+
+    private static void addSessionCookie(HttpResponse httpResponse, HttpSession httpSession) {
```
현재구조로는 Session을 사용하지 않아도 Cookie에 값이 담겨질것 같아요. 
세션을 사용한 경우에만 Session을 전달할수 있을까요 ?

sah3122 commented:
이번 미션은 간단하지만, 고민할부분이 많은 미션이라고 생각해요. 

미션의 요구사항을 정리하면, 아래와 같습니다. 
HttpRequest에서 Session을 가져올때 세션이 있다면, 기존 세션을 가져와야 하고, 없다면 새로 생성해주어야 합니다. 여기까지 구현은 잘해주셨어요 😄 
다음 스탭으로 Response 에 SessionId 를 실어주어야 하는데, 현재 구현해주신 방법은 세션 사용 유무와는 관계 없이 SessionId를 싣고 있는것 같아 코멘트 남겨드려요 😄 

Tomcat 에서 Session 구현을 위해 Response 객체를 어떻게 다루었는지 한번 보시면 좋을것 같아 링크 남겨드려요. 
https://github.com/apache/tomcat/blob/6916f850b0075513aea68b762d8aa530b3b5350b/java/org/apache/catalina/connector/Request.java#L677-L684

톰캣에서는 Request 객체가 Response 객체를 상태값으로 가지고, 세션 사용시점에 Response Cookie 에 값을 할당해줄수 있는 구조로 만들었더라구요. 이와 같은 구현은 참고만 해주시면 될것 같습니다. !

sah3122 reviewed (changes requested):
깔끔한 세션 구현 좋네요 ㅎㅎ 💯 
간단한 생각거리 남겨두었는데 확인 부탁드려요 !

ohminho-woowahan committed:
refactor
- 세션을 사용할 때 세션을 초기화 하도록 수정
- HttpRequest 에서 Session 관련 코드 제거
- HttpSession 이 필요한 RequestHandler 에 HttpSessionHandler 주입

5minho commented:
> 이번 미션은 간단하지만, 고민할부분이 많은 미션이라고 생각해요.
> 
> 미션의 요구사항을 정리하면, 아래와 같습니다. HttpRequest에서 Session을 가져올때 세션이 있다면, 기존 세션을 가져와야 하고, 없다면 새로 생성해주어야 합니다. 여기까지 구현은 잘해주셨어요 😄 다음 스탭으로 Response 에 SessionId 를 실어주어야 하는데, 현재 구현해주신 방법은 세션 사용 유무와는 관계 없이 SessionId를 싣고 있는것 같아 코멘트 남겨드려요 😄
> 
> Tomcat 에서 Session 구현을 위해 Response 객체를 어떻게 다루었는지 한번 보시면 좋을것 같아 링크 남겨드려요. https://github.com/apache/tomcat/blob/6916f850b0075513aea68b762d8aa530b3b5350b/java/org/apache/catalina/connector/Request.java#L677-L684
> 
> 톰캣에서는 Request 객체가 Response 객체를 상태값으로 가지고, 세션 사용시점에 Response Cookie 에 값을 할당해줄수 있는 구조로 만들었더라구요. 이와 같은 구현은 참고만 해주시면 될것 같습니다. !

리뷰어님! 피드백 감사합니다 
말씀해주신 대로 세션을 사용하는 시점에 세션을 생성하도록 바꾸었습니다.
톰캣처럼 HttpRequest 에서 HttpResponse 의 참조를 가지고 있으면서 처리할 수도 있겠지만.. 
저는 HttpRequest 가 HttpResponse 서로 모르는 것이 유지보수에 좋다고 생각을 했습니다.

그래서 톰캣처럼 구현하지 않고, 세션을 다루는 HttpSessionHandler 라는 객체를 만들어, 세션을 사용해야하는 RequestHandler 에 주입하는 식으로 구현해봤습니다.
HttpSessionHandler 에서 세션을 가져오는 시점에 request 에 세션 쿠키 유무를 검사해 세션을 새로 생성 또는 세션 스토어에서 가져오도록 해봤습니다.  
감사합니다.

sah3122 commented on src/main/java/webserver/JwpSessionHandler.java:
```diff
@@ -0,0 +1,47 @@
+import webserver.http.HttpResponse;
+import webserver.http.HttpSession;
+
+public class JwpSessionHandler implements HttpSessionHandler {
```
JwpSessionHandler 정의 👍

sah3122 commented on src/main/java/webserver/JwpSessionHandler.java:
```diff
@@ -0,0 +1,47 @@
+        return sessionStore.getSession(sessionIdFromCookie);
+    }
+
+    private void addSessionCookie(HttpResponse httpResponse, HttpSession httpSession) {
```
Request, Response의 불필요한 결합을 줄여주셨네요 👍 
저도 마찬가지로 Request 와 Response 는 서로 결합이 없었으면 좋을것 같다고 생각합니다. 😄

sah3122 reviewed (approved):
피드백 반영 잘해주셨네요 👍 
마지막 미션까지 화이팅입니다 😄
